core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -808,8 +808,15 @@ public MergeWriterNode plan(Merge merge)
                     assignments.put(field, rewritten);
                 }
                 else {
-                    rowBuilder.add(field.toSymbolReference());
-                    assignments.putIdentity(field);
+                    Expression expression = field.toSymbolReference();
+                    if (mergeCase instanceof MergeInsert && nonNullableColumnHandles.contains(dataColumnHandle)) {
+                        ColumnSchema columnSchema = dataColumnSchemas.get(fieldNumber);
+                        String columnName = columnSchema.getName();
+                        expression = new Coalesce(expression, new Cast(failFunction(metadata, INVALID_ARGUMENTS, "Assigning NULL to non-null MERGE target table column " + columnName), columnSchema.getType()));
```

Comment:
The error code should be `CONSTRAINT_VIOLATION`, not `INVALID_ARGUMENTS`. Also, change the message to: 'NULL value not allowed for NOT NULL column: <column>"